### ISSUE TITLE
defaultCellRangeRenderer should allow caching cells for components wrapped by ArrowKeyStepper for performance

### ISSUE THREAD

**Issue by pterodragon** [bvaughn/react-virtualized#799]
I'm fiddling around with the `ArrowKeyStepper` example and I realized that there is a perceivable lag when
- the table is modified to be 1000*1000 with height 500
- arrow down key is "keyDown", i.e. when I press and keep the arrow down key down until it reaches the bottom of the table

After digging into it for some time I know that
- visible cells + possibly some other overscan cells are only cached when the table is scrolling (by mouse wheel etc.) and not when the arrow key is down

Well, in this case I know that the cells are deterministic and static so it is unnecessary to rerender the cells every time the selected cell is moved

Then I made the following quick hacky change to solve the performance issue
- pass `scrollToRow`, `scrollToColumn`, `prevScrollToRow`, `prefScrollToColumn` from the `ArrowKeyStepper` to the Grid
- make Grid pass `scrollToRow`, `scrollToColumn`, `prevScrollToRow`, `prefScrollToColumn` to the `cellRangeRenderer`
- make the `defaultCellRenderer` cache the cells anytime except when (`rowIndex`, `columnIndex`) is either (`scrollToRow`, `scrollToColumn`) or (`prevScrollToRow`, `prevScrollToColumn`)

I know that I can provide my own `cellRenderer`/`cellRangeRenderer` but I just want to point out that the default has such performance issue. Or am I supposed to achieve the same thing using some other ways?

### Commit of concern
Basically the latest commit of master so far
```
commit 6cf3e44a8aa024bce0c804b1baecfb70e494631c
Merge: ad6837f 8eac863
Author: Brian Vaughn <brian.david.vaughn@gmail.com>
Date:   Thu Aug 31 14:27:23 2017 -0700

    Merge pull request #796 from tcosentino/clear-all-counts
    
    Clear row/column counts with CMC.ClearAll
```

### Other environment details
Browser used: Firefox 55.0.2 (64-bit)

**Comment by bvaughn** [comment]
Thanks for taking the time to write such a detailed issue.

My initial thoughts are:
* This seems pretty edge-case to me and I'm not sure it warrants the extra complexity proposed.
* The only thing that's _cached_ when scrolling is the call to the `cellRenderer` function. I added caching in the hopes that it might provide a small perf boost but I wouldn't expect it to be _that_ significant.
* Caching cells when triggered by `ArrowKeyStepper` would prevent any active-cell-highlight a user tried to implement.

PS. What does commit 6cf3e44 have to do with this? (It's mentioned in your issue.)

**Comment by pterodragon** [comment]
Sorry if I made it confusing but I put 6cf3e44 there to specify the revision I used to reproduce the issue only

> Caching cells when triggered by `ArrowKeyStepper` would prevent any active-cell-highlight a user tried to implement

That's why `ArrowKeyStepper` just has to invalidate the previous highlighted and current highlighted cell (+ some other cases?) when arrow key is pressed in this case

I thought "scrolling" using arrow key/mouse should exhibit the same behavior but I just noticed delay when using arrow down key to scroll one row down in this case. Also I'm not sure what the extent of edge case is here... Do you mean using a large table + static+deterministic content with the `ArrowKeyStepper` is an edge case?

**Comment by bvaughn** [comment]
> Sorry if I made it confusing but I put 6cf3e44 there to specify the revision I used to reproduce the issue only

Thanks for clarifying. 😄 

> Also I'm not sure what the extent of edge case is here...

I mean the use-case you described initially:
> arrow down key is "keyDown", i.e. when I press and keep the arrow down key down until it reaches the bottom of the table

Pressing and holding an arrow key to cycle through many rows in a table/list/whatever seems like an edge-case to me. Particularly in combination with a performance problem. (For example, you can do this on the [`ArrowKeyStepper` demo page](https://bvaughn.github.io/react-virtualized/#/components/ArrowKeyStepper) without any perf issues that I'm noticing.)

**Comment by pterodragon** [comment]
Alright, I guess to fine-tune in this case, users are supposed to supply their own renderers using the given interface then

**Comment by bvaughn** [comment]
Thanks for being so understand. Much appreciated~
